redis.conf

```diff
@@ -1697,6 +1697,30 @@ slowlog-max-len 128
 # "CONFIG SET latency-monitor-threshold <milliseconds>" if needed.
 latency-monitor-threshold 0
 
+################################ LATENCY STATS ##############################
+
+# The Redis extended latency stats track per command latencies and enable
+# exporting the percentile distribution and cumulative latency distributions,
+# via the INFO LATENCYSTATS command.
+# We have two groups of outputs on INFO LATENCYSTATS:
+#   - latency by percentile distribution.
+#   - cumulative distribution of latencies.
+# Using the cumulative distribution of latencies you can merge several stats
+# from different nodes to calculate aggregate metrics. Note that percentile
+# distribution is not mergeable between nodes.
+#
+# By default, the extended latency monitoring is enabled since the overhead
+# of keeping track of the command latency is very small.
+# if you don't want to track extended latency metrics, you can easily disable it
+# at runtime using the command:
+# "CONFIG SET latency-track no" if needed.
+# latency-track yes
+#
+# By default the exported latency percentiles are the p50, p99, and p999.
+# You can alter them at runtime using the command:
+# CONFIG SET latency-track-percentiles "0.0 50.0 100.0"
+# latency-track-percentiles 50.0 99.0 99.9
```

Comment:
I feel like we would save a lot of trouble by just having this be a string config that was comma separated instead of space separated. We could keep the string stored in the server struct for rewrite and get, it's not that many bytes, and on set and load we could parse the string and set it.

Reply:
we already have other mechanisms we should follow, like `client-output-buffer-limit`, `save`, and `oom-score-adj-values`. i think we should mimic these (all use `MULTI_ARG_CONFIG`)

Reply:
@oranagra what is the argument though? They are more annoying to orchestrate around since they have to be applied specially. We don't have to stick with a bad convention.

Reply:
it's no longer annoying.. it's is now supported by the infrastructure (it can split space separated strings from the config file).
it would be ugly that we have several ways to set configs that require multiple arguments (one using a generic infrastructure we created, and one that bypasses it).
i would even consider the one that bypasses the infra, an ugly hack.

storing a string in the server struct and parsing it at runtime is not an option IMHO.
and storing both the string and the parsed array in the server struct is redundant.
i rather re-compose the string from the array when needed, than keep two copies of the same data.

Reply:
I meant it's annoying for users, every other parameter is a key/value pair except the couple of multi-arg ones. Some clients don't support it, and there is an ugly hack in our system to handle them since the rest of the system is a key/value. I expect other systems have ugly hacks as well.

I'm pretty sure the amount of code would be less if we kept around the string representation, since we don't have to do all the work to reconstruct the values and special printing. I'm not sure duplicate value is all that important, we're already spending a lot of memory on this feature.

There is also another case of this in bio_cpulist, which does basically what I'm suggesting.

Reply:
`bio_cpulist` is an ugly hack.. 
config file parsing won't even detect a syntax error there. (probably crash on assertion or exit later on).
which is arguably ok since it's an immutable config.

clients should not have any problem with multi-arg configs, since we take a single string and split it.
so from their perspective, both the options we're debating are the same (except for the separator char maybe)

Reply:
The separator character is the only difference, and it's important because the start up arguments are parsed with the same delimiter.

Sure, bio_cpulist should probably have validation, but I don't think that explicitly makes it a hack.

Reply:
not sure i understand your argument about startup args delimiter.. maybe give an example.

the trick in bio_cpulist is much less severe because that's an immutable argument.

anyway, i feel strongly that we have an infrastructure, and several configs that use it, and that's what we should aim for...
IIRC the cpulist feature was a PR by an "outsider" that was merged without much of a review.

if we still disagree, i guess we need a tie breaker. @yossigo @yoav-steinberg ?

Reply:
Sure, for start up args you need to pass in the start up arguments as individual values like `redis-server --tracking-percentiles 1 2 3` compared to the config set which takes it as a single space delimited string `config set "1 2 3"`. We had special code that assumed that all configs were key/value pairs, which works for the config set case but breaks since if you try to do `redis-server --tracking "1 2 3"` it will break.

So let me just clarify my two points:
1. I don't like multi-arg configs, I think they are hacky and the infrastructure around them was more for getting all the code to be generic more so than they intrinsically make sense. I think all configs should be key/value pairs. I dislike space as a delimiter since that is also what is used for the config file to split the name of the config.
2. I think just keeping a copy around of the string representation of a config you serialize into another form isn't that big of a deal. It simplifies the get/rewrite code path as you don't have to deserialize the value, and fits in with the "apply function" that is implemented. Requirepass is an example of another config that does something similar, we transform it into a hash when we set it, but we keep it around the plaintext version in the server struct to easily do rewrite/get.

I don't really feel strongly about it, but I also don't see the motivation behind it.

Reply:
ohh. startup arguments by CLI... i didn't think of that..
i think it should be the right approach to have `redis-server --tracking "1 2 3"`, and not `redis-server --tracking 1 2 3`.
and i was guessing that this would implicitly work (with the new infrastructure).

my only concern is that if currently `redis-server --save "3601 1 301 101"` is not working while `redis-server --save 3601 1 301 101` is working, then i wouldn't want to make a breaking change.

but testing Redis 5.0, 6.0, and 6.2 shows that neither of the methods is working.
while on unstable, the later (without `"`) is working, and the first (with `"` is not working.
i think we should fix that, and considering it wasn't working at all in any of the released versions, that won't be a breaking change.

@yoav-steinberg please confirm

Reply:
Looking at the current implementation it seems like it'll work without merging the config parameters into a single arg (without adding `"`). I can change this if we want, but maybe there's no need to. WDYT?

Reply:
then how can it process the next config? by looking for `--`? what if some config has `--` as part of its value?

first let's confirm that this was never working in any of the released versions (i.e. not a regression of 6.2 or 6.0).
then i think that it'll make sense to use `"` as that's a single argument as a single config.
i.e. same as CONFIG SET takes one bulk argument as value, and not multiple bulks

Reply:
> then how can it process the next config? by looking for --? what if some config has -- as part of its value?

The current code looks for an arguments that starts with `--` and then assumes it's a config name. Then all arguments after it (that don't start with `--`) are parameters to that config. If a parameter to a config starts with `--` then we're screwed.

I'll check the old versions.

Reply:
In both 6.0 and 6.2 the logic is the same as today a CLI config argument receives it's parameters as multiple command line arguments as long as it doesn't encounter a `--`. So I think we can't break the existing logic.

Reply:
two questions:
1. what if the first argument is using `--`? e.g. `redis-server --loglevel --rdbcompression no`
2. maybe for MULTI_ARG configs, we can still break a single argument into several if we find spaces (seems consistent with CONFIG command).

assuming 1 is ok, then we can let users work around that limitation of searching for `--`, by breaking with spaces.
e.g. `redis-server --dbfilename "--odd folder name--" --loglevel verbose` will work (as it does today), but also, if there was a similar multi-arg config that could expect `--` in input it'll work `redis-server --replicaof "--no-- --one--" --loglevel verbose`
i.e. `replicaof` is a multi-arg config, and it'll be able to take arguments that it can't today.

if we do that, both `redis-server --save 20 1 --loglevel verbose` and `redis-server --save "20 1" --loglevel verbose` will work.

Reply:
@yoav-steinberg what did we conclude here?
This is unrelated to the current PR, but i see an opportunity to improve redis 7.
in past redis versions, you could only do `src/redis-server --save 1 1 --save 2 2`
now you can also do `src/redis-server --save 1 1 2 2`
but you still can't do `src/redis-server --save "1 1 2 2" (although config set does take a single arg like "1 1 2 2").
and you can't also do `src/redis-server --dbfilename --myfilename.rdb`
i think these are two things that would be nice to fix in 7.0

Reply:
made a PR to try that: https://github.com/redis/redis/pull/10042
we can continue the discussion there.

but either way, i'm confident that the new config here (`latency-track-percentiles`) should behave the same as other configs we have (e.g. `save`)